### ISSUE TITLE
Regarding GCP part of this particular project.

### ISSUE THREAD

**Issue by rishyjee** [groovenauts/iost-demo#18]
Hello everyone, I am facing some difficulty in how to integrate this project with GCP and which are the required things which we must execute on Raspberry Pi, though there are many Readme docs which are illustrated for the same but I am not able to get it clearly. 
Can anyone please help me with the GCP configuration and Raspberry Pi too. 

Thank you so much.

**Comment by rishyjee** [comment]
gs://gcp-iost-models/object_detection/20180525/saved_model/ Object Detection
gs://gcp-iost-models/object_detection/20180525/saved_model/ Cart prediction
 
I am not able to get these pre-trained models. Could you please assist me with that.

**Comment by nagachika** [comment]
Hello rishyjee,
Thank you for your attention.

I have confirmed that these files are public in Google Cloud Storage.
Try again with the links below.

(Object Detection)
https://storage.googleapis.com/gcp-iost-models/object_detection/20180525/saved_model/saved_model.pb
(Cart Prediction)
https://storage.googleapis.com/gcp-iost-models/cart_prediction/20180522/saved_model/saved_model.pb

Then upload these files to your own GCS bucket and change `GCS_URL_OF_SAVED_MODEL` in [the command line](https://github.com/groovenauts/SmartShoppingNavigator/blob/master/ml/README.md) to point directory where you upload file.

I will fix links in ml/README.md.

**Comment by rishyjee** [comment]
Okay fine, Thank you so much nagachika, for your prompt reply. 
let me try this one, once I am done, I will let you know thanks.

**Comment by rishyjee** [comment]
[image: image.png]
Still, I am getting this error, though I have linked this project with one
billing account.
[image: image.png]
I want to continue this project but still I am getting this error.
[image: image.png]
Could you please help me with this, I am not able to complete it at every
stage. And also I am not able to upload the live images from Raspberry Pi
as well.

**Comment by rishyjee** [comment]
Now it is working fine, but Now I am not able to get further what to be
done next.
I have made changes in my deployment.yaml.example file as well. I am not
getting that where to upload that ml model of model.py and What to be done
by that misc file as well.

Please do respond me as soon as possible. I shall really be obliged and
thankful to you.

Regards,
rishyjee.

**Comment by nagachika** [comment]
Hi rishyjee,

Did you deploy the pre-trained model to Cloud ML Engine in your GCP project?
https://github.com/groovenauts/SmartShoppingNavigator/tree/master/ml#deploy-models-to-ml-engine

You should create GKE cluster and deploy pods onto it.
See https://github.com/groovenauts/SmartShoppingNavigator/tree/master/gke#run-on-gke
I'm afraid there're only minimal abstract instructions are documented.
In short, you should,

1. build Docker image under  `gke/` subdirectory.
2. push the image to Container Registry
3. create Cloud Pub/Sub subscription to subscribe the topic of Cloud IoT Core
4. create GKE cluster
5. deploy pods with image build in 2. onto the cluster (with appropriate deployment.yml)

Regards,

**Comment by nagachika** [comment]
The Cloud Pub/Sub subscription required the topic to subscribe, and the topic is made from Cloud IoT Core.
SmartShoppingNavigator use Cloud IoT Core to communicate with RasPi devices.

Please see the following link.
https://github.com/groovenauts/SmartShoppingNavigator/tree/master/raspberrypi#create-device-in-cloud-iot-core

And please ignore the files under `misc/` subdirectory. These are tools chains to prepare data for model taining and dashboard etc.
You don't have to use them if you don't want to customize the food stuff and the recipes.

**Comment by rishyjee** [comment]
First of all, Thank you so much for replying and that too with so
much information in detail. So, when I am running this one for GKE cluster
then I am getting this prompt of uses exactly one argument.
[image: image.png]
And when I using kubectl then I am getting this error
[image: image.png]
And one more thing does the contents of the file deployment.yml is same as
deployment.yml.example only the values has to be changed. Correct???

I am really sorry for asking so many things multiple times, Actually, I am
not able to fix the things instantly. If it is possible then could you
please send me screencast video of the same. I mean how to configure
everything both in the RasPi and in the GCP then it would be very much
clear and helpful for me.

Thank you so much.

Regards,
rishyjee.

**Comment by nagachika** [comment]
Hi rishyjee,

I cannot see the screenshots you uploaded(?).
Could you please show your command line and the error messages in text?

> And one more thing does the contents of the file deployment.yml is same as deployment.yml.example only the values has to be changed. Correct??? 

You are right.
You should replace the values for `spec.template.spec.containers.0.env` and image name.
In other words, you should replace the following words in `deployment.yml.example`: `MY-PROJECT-ID`, `MY-SUBSCRIPTION`, `my-bucket`, `model_name`, `registry_name`.

**Comment by rishyjee** [comment]
Yeah sure,
First error
rishyjee@cloudshell:~ (smartshoppingnavigator-220720)$ docker build -t
gcr.io/smartshoppingnavigator-220720/iost-worker"docker build" requires
exactly 1 argument.See 'docker build --help'.Usage: docker build [OPTIONS]
PATH | URL | - [flags]

Build an image from a Dockerfile
The Second error is
rishyjee@cloudshell:~/SmartShoppingNavigator/gke
(smartshoppingnavigator-220720)$ kubectl create -f deployment.ymlerror:
error validating "deployment.yml": error validating data:
ValidationError(Deployment.spec): missing required field "selector" in
io.k8s.api.apps.v1beta2.DeploymentSpec; if you choose to ignore these
errors, turn validation off with --validate=false

**Comment by rishyjee** [comment]
I mean, you have been really supportive and you are giving me time to
explain everything, for that I am really grateful to you.

And one more thing, if I am not able to communicate the thing properly,
then can we have remote access of the system at 9.30 IST ie. after 1 and
half hour later. Or could you please send a screen recorded video of all of
these configurations.

I shall really be obliged to you.

Thank you.

Regards,
rishyjee.

**Comment by nagachika** [comment]
> rishyjee@cloudshell:~ (smartshoppingnavigator-220720)$ docker build -t gcr.io/smartshoppingnavigator-220720/iost-worker
> "docker build" requires exactly 1 argument.See 'docker build --help'.Usage: docker build [OPTIONS] PATH | URL | - [flags] 

You missed the last argument `.` in the command line. The correct command line is `docker build -t gcr.io/smartshoppingnavigator-220720/iost-worker .`.  Note that the last `.` indicate `current directory`.

> (smartshoppingnavigator-220720)$ kubectl create -f deployment.yml
> error: error validating "deployment.yml": error validating data: ValidationError(Deployment.spec): missing required field "selector" in io.k8s.api.apps.v1beta2.DeploymentSpec; if you choose to ignore these errors, turn validation off with --validate=false

Hmm, it seems that the new version of k8s requires some fields by default.
I pushed some changes in deployment.yml.example. Try with this new template.
https://github.com/groovenauts/SmartShoppingNavigator/commit/c5cd13492cfe4268046f4fe1243e7c789b4e6e79#diff-dcc21f2db02d37a4d632a376c62c672cR5

**Comment by rishyjee** [comment]
Okay fine, let me try this one!!!

**Comment by rishyjee** [comment]
rishyjee@cloudshell:~/SmartShoppingNavigator/gke
(smartshoppingnavigator-220720)$ kubectl create -f deployment.ymlerror:
error validating "deployment.yml": error validating data:
ValidationError(Deployment.metadata): unknown field "label" in
io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta; if you choose to ignore
these errors, turn validation off with
--validate=falserishyjee@cloudshell:~/SmartShoppingNavigator/gke
(smartshoppingnavigator-220720)$Again I am getting this error regarding
label!!!

**Comment by rishyjee** [comment]
In deployment.yml for BUCKET_PREDICTION_MODEL, "model.py" or This one
"staging.smartshoppingnavigator-220720.appspot.com"

Thank you.

**Comment by nagachika** [comment]
Oh, I'm sorry for my mistake.
https://github.com/groovenauts/SmartShoppingNavigator/commit/c79736e294552ade0a8745d60593cc8489894191 will fix the k8s deployment issue.

And, you should specify [Cloud ML Model name](https://cloud.google.com/ml-engine/docs/tensorflow/managing-models-jobs#managing_models) in deployment.yml for environment variables `ML_MODEL` and `BUCKET_PREDICTION_MODEL`.

See https://github.com/groovenauts/SmartShoppingNavigator/tree/master/ml#deploy-models-to-ml-engine too. The modell name was indicated by `MODEL` in the command line template.

**Comment by rishyjee** [comment]
In GKE for RasPi it is throwing an error of bundle: command not found in
the case of bundle install --path vendor/bundle.

**Comment by rishyjee** [comment]
That problem of GKE is fixed by that new patch.

Now when I am running these commands on RasPi then I am getting this
error...

Python 3.5.3 (/usr/bin/python3)
>>> %cd /home/pi/SmartShoppingNavigator/raspberrypi/files/home/pi
>>> %Run watchdog.py
Traceback (most recent call last):
  File
"/home/pi/SmartShoppingNavigator/raspberrypi/files/home/pi/watchdog.py",
line 137, in <module>
    main(sys.argv)
  File
"/home/pi/SmartShoppingNavigator/raspberrypi/files/home/pi/watchdog.py",
line 95, in main
    _, project_id, location, registry, device, private_key = argv
ValueError: not enough values to unpack (expected 6, got 1)
>>>


This is the error which I am getting is when I am running watchdog.py on my
RasPi system.

And one more thing, it is asking for edit config.yml file.  So where is
that config.yml file in RasPi folder?

Please do reply me as soon as you get free.

Thank you so much.

**Comment by nagachika** [comment]
Did you check [raspberrypi/README.md](https://github.com/groovenauts/SmartShoppingNavigator/blob/master/raspberrypi/README.md)?
You don't have to run watchdog.py by yourself.
After the setup instructions described in README.md and reboot Raspbian, all scripts are automatically start at  a boot time.

> And one more thing, it is asking for edit config.yml file. So where is that config.yml file in RasPi folder?

No. The configuration you should setup for your projects are the files `/home/pi/ec_private.pem` and `/home/pi/watchdog.conf` in your RasPi filesystem.

https://github.com/groovenauts/SmartShoppingNavigator/blob/master/raspberrypi/files/home/pi/watchdog.conf

`watchdog.conf` should contains to specify Your Google Cloud IoT Core device resource corresponding to your RasPi device.

**Comment by rishyjee** [comment]
And one more thing, please do reply me as soon as you are free that where to check the front end ie. the output, on which link??? like in Google I/O conference they have used [https://gcp-iost.appspot.com/](url)
Then what it is in my case like from where I will be getting that link, something should be running on GCP or directly it will be notified at a particular link????

Thank you.

**Comment by nagachika** [comment]
In fact the dashboard implementation is not fully customizable.
You can deploy dashboard application to App Engine on your GCP project.
The source of GAE application is placed under `dashboard` sub-directory, but device names are had coded in the source code.
https://github.com/groovenauts/SmartShoppingNavigator/blob/master/dashboard/index.html#L26-L28

**Comment by rishyjee** [comment]
So, here, first of all, I am supposed to change that app.go directory and then I have to replace device id with my device id in app.go directory correct???
And could you please explain how to deploy dashboard application to App Engine on GCP project as well????

Thank you!!!

**Comment by nagachika** [comment]
> So, here, first of all, I am supposed to change that app.go directory and then I have to replace device id with my device id in app.go directory correct???

Yes. More precisely, please replace `picamera##` to your device ids.

> And could you please explain how to deploy dashboard application to App Engine on GCP project as well????

You can deploy via gcloud command.
If you are not familiar with App Engine, you may have to install sdk etc.
Please see documentation for details.

https://cloud.google.com/appengine/docs/standard/go/download

https://cloud.google.com/appengine/docs/standard/go/tools/uploadinganapp

In short, try
```
$ gcloud app deploy app.yaml -v v1
```
in dashboard subdirectory.
`v1` is arbitrary version name. I recommend you to specify unique version for every deployment.

**Comment by rishyjee** [comment]
Thank you so much @nagachika. But I am getting this error when I am running
my app after it's successful deployment then I am getting this error
...Failed to get setting: datastore: cannot load field "deviceId" into a
"dashboard.Setting": no such struct field

**Comment by nagachika** [comment]
Thank you for your report.
I pushed a treatment for the error.

https://github.com/groovenauts/SmartShoppingNavigator/commit/e85eaecbbf4fdb8e0b66ee9f5b25a660fdf2b9f5

Try current master.

Regard,

**Comment by rishyjee** [comment]
Thank for fixing that issue, now it's loading properly but the problem is
that it is loading the contents of gcp-iost.appspot.com only, I mean the
link is mine but the content is of gcp-iost only. It's not capturing the
images or video from my RasPi's Camera. And One more thing, am I supposed
to change const DefaultDeviceId = "picamera01"
or all the attributes should be changed with my device ID which is
"picamera01" where "deviceId" is there.

Please do respond as soon as possible. Thank you.

**Comment by nagachika** [comment]
As you mentioned, please change `DefaultDeviceId` to the deviceId of your Cloud IoT Core.

There's one more setting (I hope this is the final one) you should change to load images from your RasPi device.

Please set environment variable `IMAGE_BUCKET` in dashboard/app.yml.
https://github.com/groovenauts/SmartShoppingNavigator/blob/master/dashboard/app.yaml#L11
It should be the same bucket you set to `SAVE_BUCKET` in gke/deployment.yml.
https://github.com/groovenauts/SmartShoppingNavigator/blob/master/gke/deployment.yml.example#L23-L24

**Comment by rishyjee** [comment]
I didn't get this one which you stated earlier  There's one more setting (I
hope this is the final one) you should change to load images from your
RasPi device.???
And the same copy of the cloud should be stored on RasPi as well, Right???

**Comment by rishyjee** [comment]
And I made all the possible changes which we have discussed but the image
from RasPi is not getting sent to the cloud. I mean it's just loading only
and refreshing. Do you have anything which I am missing in these setups????

Please do reply me whenever you get time.

Thank you.

**Comment by nagachika** [comment]
I ment you should set `IMAGE_BUCKET` in `dashboard/app.yml` to be equal to the value of `SAVE_BUCKET` in `gke/deployment.yml`.

I'd like to explain the image data flow.

1. Raspberry Pi send image to Cloud IoT Core registry
2. Cloud IoT Core send data to Cloud Pub/Sub
3. GKE container (pull.rb) pull image from Cloud Pub/Sub, save it into the Google Cloud Storage bucket.
4. Raspberry Pi display & dashboard show the image from Google Cloud Storage

You can check the where's a fault.

a) Open Cloud IoT Core registry details in GCP Console
  Open URL like `https://console.cloud.google.com/iot/locations/LOCATION/registries/REGISTRY?project=GCP_PROJECT` (Please replace UPPER CASE word. For example, LOCATION=us-central1)
  Check `Last seen` culumn of your device. If device (RasPi) upload image continuously, the value of the column should be a timestamp within a few minutes from the current time. 

b) Check GKE container logs
  Open `Stackdriver Logging` page (https://console.cloud.google.com/logs/viewer?project=GCP-PROJECT, Please replace GCP-PROJECT with your project id)
  Select `GKE container > cluster name` in resource dropdown list.
  Check if there're any error messages.
